From 748af3c1737a15afb10cd3df8fe178ffc087d20f Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Mon, 23 Jun 2025 16:27:04 +0800
Subject: [PATCH 06/63] loonggpu: Improve fbdev object-test helper

Follow the change of our reference implementation, allowing us to
remove struct gsgpu_fbdev later.

Link: https://git.kernel.org/torvalds/c/276f7b4bd524
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 loonggpu/loonggpu_fb.c | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/loonggpu/loonggpu_fb.c b/loonggpu/loonggpu_fb.c
index 1f82c8e..7141563 100644
--- a/loonggpu/loonggpu_fb.c
+++ b/loonggpu/loonggpu_fb.c
@@ -10,7 +10,7 @@
 #include "loonggpu_drm.h"
 
 #include <drm/drm_fb_helper.h>
-
+#include <drm/drm_gem_framebuffer_helper.h>
 
 #include "loonggpu_display.h"
 #include "loonggpu_helper.h"
@@ -338,9 +338,17 @@ void loonggpu_fbdev_set_suspend(struct loonggpu_device *adev, int state)
 
 bool loonggpu_fbdev_robj_is_fb(struct loonggpu_device *adev, struct loonggpu_bo *robj)
 {
-	if (!adev->mode_info.rfbdev)
+	struct drm_fb_helper *fb_helper = adev->ddev->fb_helper;
+	struct drm_gem_object *gobj;
+
+	if (!fb_helper)
 		return false;
-	if (robj == gem_to_loonggpu_bo(adev->mode_info.rfbdev->rfb.base.obj[0]))
-		return true;
-	return false;
+
+	gobj = drm_gem_fb_get_obj(fb_helper->fb, 0);
+	if (!gobj)
+		return false;
+	if (gobj != &robj->tbo.base)
+		return false;
+
+	return true;
 }
-- 
2.52.0

